### ISSUE TITLE
I can't set variables on a workflow instance with a cluster of multiple partitions

### ISSUE THREAD

**Issue by Zelldon** [zeebe-io/zeebe#6214]
**Describe the bug**

I want to set multiple variables on a workflow instance. This works only if I have one partition, otherwise I get.

```
Error: rpc error: code = NotFound desc = Command 'UPDATE' rejected with code 'NOT_FOUND': Expected to update variables for element with key '15762598695796736', but no such element was found
```

**To Reproduce**

```sh
#!/bin/bash

set -exuo pipefail

source CamundaCloudMgmtAPI-Client-Test.txt
zbctl deploy one_task.bpmn

key=$(zbctl create instance benchmark | jq '.workflowInstanceKey')
randString=$(tr -dc A-Za-z0-9 < /dev/urandom | head -c $((64)); echo '')
for i in {1..10}
do
  variables=$(jq -n --arg key "test-$i" \
    --arg var "$randString" \
    '.[$key] = $var')

  zbctl set variables "$key" --variables "$variables"
done
```

**Expected behavior**
It works and there is no difference whether one or multiple partition I have.

**Log/Stacktrace**
<details><summary>Full Stacktrace</summary>
 <p>

```
+ zbctl deploy oneTask.bpmn caller.bpmn
{
  "key": 2251799813685258,
  "workflows": [
    {
      "bpmnProcessId": "oneTask",
      "version": 1,
      "workflowKey": 2251799813685249,
      "resourceName": "oneTask.bpmn"
    },
    {
      "bpmnProcessId": "benchmark",
      "version": 1,
      "workflowKey": 2251799813685250,
      "resourceName": "caller.bpmn"
    }
  ]
}
++ tr -dc A-Za-z0-9
++ head -c 64
++ echo ''
+ randString=fKo2hi7hlpy8jKd7ukuC8J3zORPR3qOt0MM602nXkwHsGg7Q2GwtOhlF0rCk087J
+ for i in {1..1000}
++ zbctl create instance benchmark
++ jq .workflowInstanceKey
+ key=15762598695796736
+ for i in {1..1000}
++ jq -n --arg key test-1 --arg var fKo2hi7hlpy8jKd7ukuC8J3zORPR3qOt0MM602nXkwHsGg7Q2GwtOhlF0rCk087J '.[$key] = $var'
+ variables='{
  "test-1": "fKo2hi7hlpy8jKd7ukuC8J3zORPR3qOt0MM602nXkwHsGg7Q2GwtOhlF0rCk087J"
}'
+ zbctl set variables 15762598695796736 --variables '{
  "test-1": "fKo2hi7hlpy8jKd7ukuC8J3zORPR3qOt0MM602nXkwHsGg7Q2GwtOhlF0rCk087J"
}'
Error: rpc error: code = NotFound desc = Command 'UPDATE' rejected with code 'NOT_FOUND': Expected to update variables for element with key '15762598695796736', but no such element was found
```

</p>
</details>

**Environment:**
 - OS: cloud, production l
 - Zeebe Version: 0.26.0
 - Configuration: [e.g. exporters etc.]

**Comment by npepinpe** [comment]
Can you update any other variable scope (e.g. activity) once this happens, and would this also occur again?

**Comment by npepinpe** [comment]
@korthout as the medic can you clarify the impact/scope here? And if along the way you root cause it, all the better!

**Comment by korthout** [comment]
tl;dr:  we don't always communicate the correct workflow instance key to the client in a multi-partition cluster.

I was able to reproduce it not just once, but twice. And then, not anymore... Even deleting my cluster (Production-M) and starting out fresh did not make it reappear. I had no clue what changed... But then I found a way to reproduce it again. Running this script more often. 

Somehow, once in a while the first set variable command fails with this message. Note that subsequent set variable commands for the same workflow instance always succeed or always fail no matter how often they're tried. The message itself is produced by [UpdateVariableDocumentProcessor.java#L45](https://github.com/zeebe-io/zeebe/blob/develop/engine/src/main/java/io/zeebe/engine/processing/variable/UpdateVariableDocumentProcessor.java#L45). So the variable scope element (which in the example is the workflow instance) cannot be found in the state. It seemed then that this is the wrong partition.

Because the processor is processing it, we know that the UpdateVariableDocumentRecord was written to a stream (of a specific partition). The gateway determines which partition using `Protocol.decodePartitionId(elementInstanceKey)`. The elementInstanceKey in this case is the workflow instance key. 

So I decoded the workflow instance key for some of the successful and failed commands: 

```
/* key => partitionId (success) */
9007199254741016 => 4 (failed)
9007199254741012 => 4 (failed)
6755399441055807 => 3 (success)
2251799813685375 => 1 (success)
4503599627370581 => 2 (success)
4503599627370607 => 2 (success)
2251799813685401 => 1 (success)
6755399441055851 => 3 (success)
9007199254741024 => 4 (failed)
4503599627370619 => 2 (success)
6755399441055877 => 3 (success)
2251799813685419 => 1 (success)
9007199254741028 => 4 (failed)
```
Note that all failures occur on the partition 4.

So I went to look for these failed workflow instance in operate:
```
9007199254741012
9007199254741016
9007199254741024
9007199254741028
```
And they're not there (there's a couple more instances, because I did not collect all in the above table)
![Screenshot 2021-01-28 at 22 17 53](https://user-images.githubusercontent.com/3511026/106200104-bdbaa500-61b6-11eb-80af-bbd8f13d74f7.png). 

For some reason, these workflow instances don't exist, at all.

I emptied the cluster, reran the test. 3 runs were successful and the fourth failed. It failed with workflow instance key 9007199254741048. 
![Screenshot 2021-01-28 at 22 22 40](https://user-images.githubusercontent.com/3511026/106200489-594c1580-61b7-11eb-8937-f50dc92fb8a4.png)

It seems we don't always communicate the correct workflow instance key to the client. This seems quite severe to me.

**Comment by menski** [comment]
Hey guys, not sure if I'm correct but at least for me this looks like a bug in `jq` or at least a behavior I cannot explain. When I checked Operate the workflow instance key for me was always one higher then the one returned by `jq`. So I added a `tee` in the pipe to check the real output and so that happend:

```
$ zbctl create instance single-task | tee output.json |  jq '.workflowInstanceKey'
9007199254741016

$ cat output.json 
{
  "workflowKey": 2251799813685249,
  "bpmnProcessId": "single-task",
  "version": 1,
  "workflowInstanceKey": 9007199254741017
}
```

So the following looks wrong to me

```
$ echo '{"foo": 9007199254741005}' | jq '.foo'
9007199254741004
```

Can you please verify that this is not a problem of the script but really Zeebe?

**Comment by Zelldon** [comment]
wtf @menski you are right :exploding_head: I found this

> In short, jq uses IEEE754 doubles to store numbers (which is permitted by the JSON specification). This means that very large integers might get adjusted to the nearest representable value. (Even if you weren't using jq, it's possible some other tool you use would do this to you)
> https://github.com/stedolan/jq/issues/1959#issuecomment-519659330

**Comment by Zelldon** [comment]
@korthout I think the workflow instance doesn't exist because the key was translated, as @menski mentioned:

See

```
[zell ~/ cluster: zeebe-cluster ns:default]$ echo '{"foo": 9007199254741023}' | jq '.foo'
9007199254741024
[zell ~/ cluster: zeebe-cluster ns:default]$ echo '{"foo": 9007199254741025}' | jq '.foo'
9007199254741024
```

**Comment by Zelldon** [comment]
I run an IT test with multiple partitions and set variables and this worked. I think we can close this. Sorry for any inconvenience, wasn't aware of this behavior.

P.S.: we can improve our IT test coverage for multiple partitions :)

**Comment by korthout** [comment]
@menski @Zelldon Thanks for looking into this
